### DISCUSSION TITLE
ESP32 port

### DISCUSSION THREAD

**Discussion by DeeEmm** [DIY-Flow-Bench/DIY-Flow-Bench#49] [Code Development]
![UNO-ESP325-650x350](https://user-images.githubusercontent.com/3038710/126052436-e4eea9ec-64f7-4d7a-b56b-e3f4df5cdbe1.jpg)

I've been playing with ESP32 processors on a few other projects and think that it might be a good fit for the DIYFB project.

The ESP32 is sold as an all-in-one IOT style processor which bundles a lot of features into one package. Most notably it includes on-board bluetooth and WiFi connectivity which simplifies peripheral equipment connectivity as device control can be made via a simple browser application.

The ESP is also a 32 bit processor compared to the Atmel which is only 8 bit. This means that ADC resolution is much better, which in turn means more accurate results. Processor power is also massively improved, not that we are really pushing the limits of the Arduino, but we do avoid using floating point math to prevent overloading things.  

The ESP can also be programmed via the Arduino IDE so can still be accessible to those familiar with the IDE.

I've ordered an ESP32 that shares the Uno footprint with the intention of making the DIYFB shield pin compatible with the Arduino Uno. This will allow people to use either processor should they desire.

The inclusion of the WiFi / and web browser support means that I will most likely drop the support for external displays on the ESP32 and instead rely on the browser and serial API. TCMenu support for the Arduino will still be retained however.

I will update here and the github repo with progress once I have received the ESP32 and made some headway

**Comment by DeeEmm** [comment]
OK. So I've created a branch for the ESP32 port. 

The plan is to port the existing code so that it works on the ESP32 platform and go from there.

**Comment by DeeEmm** [comment]
UPDATE: I have ported the existing code across and have it running on the ESP32. I have also created a basic web server and web interface to display measurement data.

<img width="1016" alt="Screen Shot 2021-07-19 at 11 19 28 am" src="https://user-images.githubusercontent.com/3038710/126091666-33bf415b-cd2f-4015-b568-343fee437dcc.png">

I still have a bit of work to do to get the data updating live via websockets, I also want to add in some configuration options for setting up the bench options and WiFi.

The plan is that the controller will attempt to establish a WiFi connection to the last WiFi network it was connected to, if not it will default into access point mode and can be accessed directly. This will allow you to update the wifi credentials and then connect as a client to the specified Wifi Network.

The IP address will be reported via the serial link, but I will also add a DNS translator so that you can connect via hostname. i.e. DIY_FB.local.

**Comment by DeeEmm** [comment]
Here's a demo...

https://diyflowbench.com/demo/

**Comment by jwestes** [comment]
This is good stuff.  I need to get us caught up in the shop so I can start to contribute again.

Jason Estes
***@***.***
________________________________

**Comment by DeeEmm** [comment]
<img width="1608" alt="WebUIDashboard" src="https://user-images.githubusercontent.com/3038710/126745200-fead302f-e0aa-48b6-8165-f72902b1edac.png">

Build 21072301 includes support for Multicast DNS as well as a configuration interface and a default fallback to AP mode if no WiFi connection is made. It should be possible to connect in AP mode, update the WiFi configuration settings and then reboot the device to connect to your Wifi Network.

If anyone wants to build this it is necessary to firstly upload the ESP code and then the contents of the Data folder. I've updated the [WIKI](https://github.com/DeeEmm/DIY-Flow-Bench/wiki/3.-Installation) to include some basic instructions on how to do this.

<img width="1607" alt="WebUIConfig" src="https://user-images.githubusercontent.com/3038710/126745230-22963a0e-6e29-4838-b356-41c21c210d73.png">

Configuration settings are stored in the configuration.txt file. The settings are read from this file and displayed in the Web UI. This file is also overwritten when the configuration settings in the Web UI are saved. If the file is missing it will be created the first time the settings are saved in the Web UI.

The configuration settings currently listed are operational or runtime related settings. I have avoided breaking out settings that are related to hardware as those settings will only ever need to be set once, at compile time. This includes settings like pin definitions and MAF sensor selection. 

Inclusion of other configuration parameters is not too hard to do and the list will no doubt grow as the project progresses so if anyone has any suggestions for settings that should be included, just let me know.

I have some additional things I want / need to integrate before calling the initial port done. These are outlined in the [ESP32 Port project](https://github.com/DeeEmm/DIY-Flow-Bench/projects/4) but essentially relate to calibration and being able to manage the WebUI and Configuration files via the WebUI.

The plan is essentially that with a vanilla installation you can connect to the device via AP mode (Peer-to-peer WiFi connection). At this point there are no files stored on the devices SPIFFS partition, only the firmware, so the regular WebUI is not present. Instead a basic file management interface is served direct from the ESP32 allowing the user to upload the WebUI. This is a pretty common method used by ESP32 devices. It also has the benefit that there is complete segregation between the firmware and WebUI code making changes and development easier.

Ultimately the WebUI code (HTML + CSS + JS) will be minified and rolled up into one Gzipped file as this ultimately saves disc space, which can be better utilised for storing data logs, but for the time being I will keep it in its uncompressed format.

/DM

**Comment by bikedyno** [comment]
hi DeeEmm/DIY-Flow-Bench
i have it working esp32
web interface connected as, in example
wait for more update to use new system
thanks pete

**Comment by DeeEmm** [comment]
> hi DeeEmm/DIY-Flow-Bench i have it working esp32 web interface connected as, in example wait for more update to use new system thanks pete

Thanks for the feedback Pete.

I've just committed some more changes. Mostly updates to the UI, additional configuration params and added some modal dialogs for status and file management.

Hopefully by the next commit I'll have both of those dialogs working.

/DM

**Comment by DeeEmm** [comment]
I've started a related discussion #56 on Data logging & Visual Data Representation. I would be interested to hear your opinions.

**Comment by DeeEmm** [comment]
I've broken the current ESP port tasks out into issues and added some timing to the project. I want to get the port completed in about a fortnight and the additional code features for the V2 project integrated by the month end. Alongside this I will endeavor to get the shield finished off.

The aim is to get some functional units ready along side the V2 beta for testing next month so that I can work through any bugs.

Ultimately I plan to allow a month for beta tests along with a further month for unit tests and then an official release

**Comment by DeeEmm** [comment]
All tasks related to the ESP32 port have now been addressed. The code has been merged into main and is now the main branch.

I have packaged everything up as a new release - V.2.0 RC-1

Please feel free to download the code and test it out. Please direct all comments relating to this version to the release discussion:

https://github.com/DeeEmm/DIY-Flow-Bench/discussions/86

/DM
